core/modules/loader/MWorkerList.cc

```diff
@@ -0,0 +1,424 @@
+                            std::string const& ourHostName, short ourPort) {
+    NetworkAddress address(ip, port);
+    StringElement workerList;
+    {
+        std::lock_guard<std::mutex> lock(_mapMtx);
+        if (_wListChanged || _stateListData == nullptr) {
+            _wListChanged = false;
+            /// At this time, all workers should easily fit in a single message.
+            /// TODO send multiple messages (if needed) with each having the address and range of 100 workers.
+            ///      This version is useful for testing. _stateListData becomes a vector.
+            proto::LdrMastWorkerList protoList;
+            protoList.set_workercount(_nameMap.size());
+            for (auto const& item : _nameMap ) {
+                proto::WorkerListItem* protoItem = protoList.add_worker();
+                MWorkerListItem::Ptr const& wListItem = item.second;
+                protoItem->set_wid(wListItem->getId());
+            }
+            protoList.SerializeToString(&(workerList.element));
+            LoaderMsg workerListMsg(LoaderMsg::MAST_WORKER_LIST, msgId, ourHostName, ourPort);
+            _stateListData = std::make_shared<BufferUdp>();
+            workerListMsg.appendToData(*_stateListData);
+            workerList.appendToData(*_stateListData);
+        }
+    }
+
+    _central->sendBufferTo(ip, port, *_stateListData);
+
+    // See if this worker is know.
+    MWorkerListItem::Ptr workerItem;
+    {
+        // See if this is a worker in our map
+        std::lock_guard<std::mutex> lock(_mapMtx);
+        auto iter = _ipMap.find(address);
+        if (iter != _ipMap.end()) {
+            workerItem = iter->second;
+        }
+    }
+    if (workerItem != nullptr) {
+        workerItem->sendListToWorkerInfoReceived();
+    }
+    return true;
+}
+
+
+
+
+std::pair<std::vector<MWorkerListItem::Ptr>, std::vector<MWorkerListItem::Ptr>>
+MWorkerList::getActiveInactiveWorkerLists() {
+    std::vector<MWorkerListItem::Ptr> active;
+    std::vector<MWorkerListItem::Ptr> inactive;
+    std::lock_guard<std::mutex> lck(_mapMtx);
+    for(auto const& elem : _nameMap) {
+        auto item = elem.second;
+        if (item->isActive()) {
+            active.push_back(item);
+        } else {
+            inactive.push_back(item);
+        }
+    }
+    auto pair = std::make_pair(active, inactive);
+    return pair;
+}
+
+
+// must lock _mapMtx before calling this function
+void MWorkerList::_flagListChange() {
+    _wListChanged = true;
+    // On the Master, flag each worker in the list that it needs to send an updated list to it's worker.
+    for (auto const& elem : _nameMap) {
+        auto const& item = elem.second;
+        item->flagNeedToSendList();
+    }
+}
+
+
+std::string MWorkerList::dump() const {
+    std::stringstream os;
+    os << "MWorkerList:\n";
+    {
+        std::lock_guard<std::mutex> lck(_mapMtx);
+        for (auto elem:_nameMap) {
```

Comment:
You don't want a copy constructor of each element of the map to be called here. Use tis instead:
```
for (auto&& elem:_nameMap) {
}
```
Since the method is `const` then `auto&& item` will be equivalent to `auto const&`.